include/ccf/pal/snp_ioctl6.h

```diff
@@ -0,0 +1,121 @@
+#pragma once
+
+#include "ccf/pal/attestation_sev_snp.h"
+
+#include <fcntl.h>
+#include <stdint.h>
+#include <sys/ioctl.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+// Based on the SEV-SNP ABI Spec document at
+// https://www.amd.com/system/files/TechDocs/56860.pdf
+
+/* linux kernel 6.* versions of the ioctls that talk to the PSP */
+
+namespace ccf::pal::snp::ioctl6
+{
+  constexpr auto DEVICE = "/dev/sev-guest";
+
+  // Table 20
+  struct AttestationReq
+  {
+    uint8_t report_data[snp_attestation_report_data_size];
+    uint32_t vmpl;
+    uint8_t reserved[28]; // needs to be zero
+  }; // aka snp_report_req in (linux) include/uapi/linux/sev-guest.h
+
+  // Table 23
+#pragma pack(push, 1)
+  struct AttestationResp
+  {
+    uint32_t status;
+    uint32_t report_size;
+    uint8_t reserved[0x20 - 0x8];
+    struct Attestation report;
+    uint8_t padding[64];
+    // padding to the size of SEV_SNP_REPORT_RSP_BUF_SZ (i.e., 1280 bytes)
+  };
+
+  static_assert(sizeof(AttestationResp) < 4000);
+  struct AttestationRespWrapper
+  {
+    struct AttestationResp resp;
+    uint8_t padding[4000 - sizeof(struct AttestationResp)];
+  };
+#pragma pack(pop)
+
+  struct GuestRequest
+  {
+    uint8_t msg_version; // message version number (must be non-zero)
+    AttestationReq* req_data;
+    AttestationRespWrapper* resp_wrapper;
+    uint64_t fw_err; // firmware error code on failure (see psp-sev.h)
+  };
+
+  constexpr char SEV_GUEST_IOC_TYPE = 'S';
+  constexpr int SEV_SNP_GUEST_MSG_REPORT =
+    _IOWR(SEV_GUEST_IOC_TYPE, 0x0, struct snp::ioctl6::GuestRequest);
+
+  static inline bool is_sev_snp()
+  {
+    return access(DEVICE, W_OK) == 0;
+  }
+
+  class Attestation : public AttestationInterface
+  {
+    AttestationReq req = {};
+    AttestationRespWrapper resp_wrapper = {};
+
+  public:
+    Attestation(const PlatformAttestationReportData& report_data)
+    {
+      if (report_data.data.size() <= snp_attestation_report_data_size)
+      {
+        std::copy(
+          report_data.data.begin(), report_data.data.end(), req.report_data);
+      }
+      else
+      {
+        throw std::logic_error(
+          "User-defined report data is larger than available space");
+      }
+
+      int fd = open(DEVICE, O_RDWR | O_CLOEXEC);
+      if (fd < 0)
+      {
+        throw std::logic_error(fmt::format("Failed to open \"{}\"", DEVICE));
+      }
+
+      // Documented at
+      // https://www.kernel.org/doc/html/latest/virt/coco/sev-guest.html
+      GuestRequest payload = {
+        .msg_version = 1,
+        .req_data = &req,
+        .resp_wrapper = &resp_wrapper,
+        .fw_err = 0};
+
+      int rc = ioctl(fd, SEV_SNP_GUEST_MSG_REPORT, &payload);
+      if (rc < 0)
+      {
+        CCF_APP_FAIL("IOCTL call failed: {}", strerror(errno));
+        CCF_APP_FAIL("Payload error: {}", payload.fw_err);
+        throw std::logic_error(
+          "Failed to issue ioctl SEV_SNP_GUEST_MSG_REPORT");
+      }
```

Comment:
Most of this logic to get an attestation is very similar to the corresponding one in `include/ccf/pal/snp_ioctl5.h`. It could make sense to factor it out to avoid duplicated code and make this part more concise in both files

Reply:
I see, I didn't know you wanted to drop 5.x support. Then it makes sense to keep it like this.